ITMO_FS/filters/multivariate/DISRwithMassive.py

```diff
@@ -88,47 +76,32 @@ def _fit(self, x, y):
                 "Cannot select %d features with n_features = %d" %
                 (self.n_features, self.n_features_))
 
-        free_features = np.array([], dtype='object')
+        free_features = np.array([], dtype='int')
         self.selected_features_ = generate_features(x)
-        self._vertices = np.ones(self.n_features_)
-        self._edges = np.zeros((self.n_features_, self.n_features_))
-        for i in range(self.n_features_):
-            for j in range(self.n_features_):
-                entropy_pair = entropy(list(zip(x[:, i], x[:, j])))
-                if entropy_pair != 0.:
-                    self._edges[i][j] = _chained_information(
-                        x[:, i], x[:, j], y) / entropy_pair
-
-        # TODO apply vectorize to selected_features and not arange(n_features)?
+        self._edges = pairwise_distances(x.T, x.T, lambda xi, xj: 
+            _chained_information(xi, xj, y) / (joint_entropy(xi, xj) + 1e-15))
+        np.fill_diagonal(self._edges, 0)
+    
         while len(self.selected_features_) != self.n_features:
-            min_index = np.argmin(
-                np.vectorize(
-                    lambda a: self.__count_weight(a))(
-                    self.selected_features_))
-            self._vertices[min_index] = 0
-            free_features = np.append(free_features, min_index)
+            min_index = np.argmin(np.sum(
+                self._edges[np.ix_(self.selected_features_, 
+                    self.selected_features_)], 
+                axis=0))
+            free_features = np.append(free_features, 
+                self.selected_features_[min_index])
             self.selected_features_ = np.delete(
                 self.selected_features_, min_index)
 
-        change = True
-        while change:
-            change = False
-            swap_index = (-1, -1)
-            max_difference = 0
-            for i in range(len(free_features)):
-                for j in range(len(self.selected_features_)):
-                    temp_difference = (self.__count_weight(
-                        free_features[i]) - self.__count_weight(
-                        self.selected_features_[j]))
-                    if temp_difference > max_difference:
-                        max_difference = temp_difference
-                        swap_index = (i, j)
-            if max_difference > 0:
-                change = True
-                new_selected, new_free = swap_index
-                free_features = np.append(free_features, new_free)
-                free_features = np.delete(free_features, new_selected)
-                self.selected_features_ = np.append(
-                    self.selected_features_, new_selected)
-                self.selected_features_ = np.delete(
-                    self.selected_features_, new_free)
+        while True:
```

Comment:
better add counter except infinite wait

Reply:
theres nothing to count actually, each iteration of the cycle improves the score by calculating the difference in it from swapping a pair of features and swapping the features with the biggest positive difference; this is by definition a finite process because there's always an optimal solution